mmf/trainers/base_trainer.py

```diff
@@ -242,22 +242,24 @@ def train(self):
             if self.current_epoch > self.max_epochs:
                 break
 
-            for batch in self.train_loader:
+            epoch_num_batches = len(self.train_loader)
+
+            for epoch_batch_index, batch in enumerate(self.train_loader):
                 self.profile("Batch load time")
-                self.current_iteration += 1
-                self.writer.write(self.num_updates + 1, "debug")
+
+                self._check_start_update(epoch_batch_index, epoch_num_batches)
```

Comment:
Another possible way which we should discuss is to do following:
- Run a while loop
- Inside this while loop, run the for loop over `update_frequency` iterations.
- Use `iter(self.train_loader)` to get the next batch.
- If StopIteration error is found break the while loop
- Otherwise, follow the same flow with the batch you got now
- Do forward pass for each of the `update_frequency` items and keep accumulating reports and losses in a list
- When iterations finish run backward, stack the loss and divide by total number of elements and similarly for report have stack function.


This will keep most of the trainer logic same and give you ability of `update_frequency` easily.

Reply:
Let's move this discussion to a chat channel and I'll post the resolution here.